### PR TITLE
Submission by round status detail view: status tab actions

### PR DESCRIPTION
**Pull request by tm-kn**

### CODE REVIEW

File: opentech/static_src/src/app/src/containers/StatusActions.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    closeStatusModal = () => {
+        this.setState({
+            modalVisible: false,
+            statusSelectValue: '',
+        });
+    }
+
+    handleStatusChange = () => {
```
Comment (tm-kn) [inline opentech/static_src/src/app/src/containers/StatusActions.js]: Add error handling.

File: opentech/apply/funds/serializers.py [hunk 2]
```diff
@@ -11,10 +12,29 @@
...
+            }
+
+            # Sometimes the status does not exist in the
+            # determination matrix.
+            redirect = None
+            try:
+                redirect = DeterminationCreateOrUpdateView.should_redirect(
+                    self.context['request'], instance, transition)
+            except KeyError:
+                pass
```
Comment (todd-dembrey) [inline opentech/apply/funds/serializers.py]:
```
       except KeyError:
            redirect = None
```

File: opentech/apply/funds/api_views.py [hunk 3]
```diff
@@ -85,7 +85,16 @@ def post(self, request, *args, **kwargs):
             obj.perform_transition(action, self.request.user, request=self.request)
```
Comment (todd-dembrey) [inline opentech/apply/funds/api_views.py]: Missing an redirect check here.

there should be two checks.
1. Should this have been a redirect? - if so they have stale data or trying to break things
  * we should prevent that
  * determinations are not coded into the workflow explicitly. so its a seperate step.
2. Do the transition (and report if it fails) - as it currently does

File: opentech/apply/funds/serializers.py [hunk 4]
```diff
@@ -11,10 +12,29 @@
...
+            action_dict = {
+                'value': transition,
+                'display': action
+            }
+
+            # Sometimes the status does not exist in the
+            # determination matrix.
+            redirect = None
+            try:
+                redirect = DeterminationCreateOrUpdateView.should_redirect(
```
Comment (todd-dembrey) [inline opentech/apply/funds/serializers.py]: Was it not possible to extract the common method?

File: opentech/apply/funds/api_views.py [hunk 5]
```diff
@@ -85,7 +85,16 @@ def post(self, request, *args, **kwargs):
...
         except DjangoPermissionDenied as e:
             raise PermissionDenied(str(e))
-        return Response(status=status.HTTP_200_OK)
+        obj = self.get_object()
+        serializer = SubmissionDetailSerializer(obj, context={
+            'request': request,
+        })
```
Comment (todd-dembrey) [inline opentech/apply/funds/api_views.py]: This is sort of ok, does too much work, but should be ok.

File: opentech/apply/funds/serializers.py [hunk 6]
```diff
@@ -80,11 +101,12 @@ def get_questions(self, obj):
 
 
 class SubmissionActionSerializer(serializers.ModelSerializer):
-    actions = ActionSerializer(source='*')
+    actions = ActionSerializer(source='*', read_only=True)
+    action = serializers.CharField(write_only=True)
```
Comment (todd-dembrey) [inline opentech/apply/funds/serializers.py]: Seeing as we don't use the proper create method is there any value to doing this? I would hate for someone to start posting actions and easily by passing the view. (always possible but lets not make it easy)

File: opentech/static_src/src/app/src/components/Modal/index.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        }
+    }
+
+    render() {
+        const { content, heading, onClose } = this.props;
+        return (
+            <div style={this.styles.modal}>
+                {onClose && <button onClick={onClose}>[X]</button>}
+                {heading && <h1>{heading}</h1>}
+                {content}
```
[the remark this answers is not in the document]
Comment (todd-dembrey) [reply opentech/static_src/src/app/src/components/Modal/index.js]: Also: https://github.com/reactjs/react-modal?

File: opentech/apply/funds/api_views.py [hunk 8]
```diff
@@ -85,7 +85,16 @@ def post(self, request, *args, **kwargs):
             obj.perform_transition(action, self.request.user, request=self.request)
         except DjangoPermissionDenied as e:
             raise PermissionDenied(str(e))
-        return Response(status=status.HTTP_200_OK)
+        obj = self.get_object()
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply opentech/apply/funds/api_views.py]: I tried that before using this, but was getting an error about immutable fields. The Python instance must be recreated it seems.

File: opentech/apply/funds/serializers.py (hunk 4 above)
Comment (tm-kn) [reply opentech/apply/funds/serializers.py]: Do you want me to extract should_redirect as a common method?

File: opentech/static_src/src/app/src/components/Modal/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (tm-kn) [reply opentech/static_src/src/app/src/components/Modal/index.js]: It's up to you and @chris-lawton whether you want to use third-party modal or not.

File: opentech/static_src/src/app/src/containers/StatusActions.js [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                return redirect(actionObject.target)
+            case 'submit':
+                return changeStatus(statusSelectValue)
+                        .then(() => {
+                            this.closeStatusModal()
+                            addMessage(
+                                'You have successfully updated the status of the submission',
+                                MESSAGE_TYPES.SUCCESS,
+                            );
+                        })
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply opentech/static_src/src/app/src/containers/StatusActions.js]: Not sure if we want an error to be a notification or a paragraph? We could add error there.

[the remark this answers is not in the document]
Comment (tm-kn) [reply opentech/static_src/src/app/src/containers/StatusActions.js]: We also can trigger a new message in redux if you want to ;)

File: opentech/apply/funds/api_views.py (hunk 3 above)
Comment (tm-kn) [reply opentech/apply/funds/api_views.py]: Updated.

File: opentech/apply/funds/api_views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (todd-dembrey) [reply opentech/apply/funds/api_views.py]: :+1: Thats my fault. limitation of FSM, see bottom of this section. https://github.com/viewflow/django-fsm/blob/5293035481b1066c510a38b59ce307ff29693ec8/README.rst#usage

File: opentech/apply/funds/serializers.py (hunk 4 above)
Comment (todd-dembrey) [reply opentech/apply/funds/serializers.py]: Lets leave it for now. It makes sense as it is. If we make the determination view API driven then we will address that then.

File: opentech/static_src/src/app/src/containers/StatusActions.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (todd-dembrey) [reply opentech/static_src/src/app/src/containers/StatusActions.js]: I think we should add it in the modal as an error message.

File: opentech/static_src/src/app/src/containers/StatusActions.js [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+        if (submission === undefined || submission.actions === undefined) {
+            return null;
+        } else if (submission.actions.length === 0) {
+            return <button disabled={true}>Update status (no actions)</button>
+        }
+
+        return (
+            <>
+                <Modal isOpen={this.state.modalVisible}
```
Comment (todd-dembrey) [inline opentech/static_src/src/app/src/containers/StatusActions.js]: It is beautiful!

File: opentech/static_src/src/app/src/containers/StatusActions.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (tm-kn) [reply opentech/static_src/src/app/src/containers/StatusActions.js]: Yes.

**Pull request merged**